nums/core/array/application.py

```diff
@@ -332,28 +334,45 @@ def diag(self, X: BlockArray) -> BlockArray:
                 syskwargs = {"grid_entry": grid_entry, "grid_shape": grid.grid_shape}
                 if np.all(np.diff(grid_entry) == 0):
                     # This is a diagonal block.
-                    rarr.blocks[grid_entry].oid = self.cm.diag(X.blocks[grid_entry[0]].oid,
+                    rarr.blocks[grid_entry].oid = self.cm.diag(X.blocks[grid_entry[0]].oid, 0,
                                                                syskwargs=syskwargs)
                 else:
                     rarr.blocks[grid_entry].oid = self.cm.new_block("zeros",
                                                                     grid_entry,
                                                                     grid_meta,
                                                                     syskwargs=syskwargs)
         elif len(X.shape) == 2:
-            assert X.shape[0] == X.shape[1], "X must be a square array."
-            assert X.block_shape[0] == X.block_shape[1], "block_shape must be square."
-            shape = X.shape[0],
-            block_shape = X.block_shape[0],
+            shape = min(X.shape),
+            block_shape = min(X.block_shape),
             grid = ArrayGrid(shape, block_shape, X.dtype.__name__)
             rarr = BlockArray(grid, self.cm)
-            for grid_entry in X.grid.get_entry_iterator():
-                out_grid_entry = grid_entry[:1]
-                out_grid_shape = grid.grid_shape[:1]
-                syskwargs = {"grid_entry": out_grid_entry, "grid_shape": out_grid_shape}
-                if np.all(np.diff(grid_entry) == 0):
-                    # This is a diagonal block.
-                    rarr.blocks[out_grid_entry].oid = self.cm.diag(X.blocks[grid_entry].oid,
-                                                                   syskwargs=syskwargs)
+            diagonal, blocks = array_utils.find_output_blocks(X.blocks, shape, block_shape)
+
+            all_result_oids = {}
+            out_grid_shape, count = (len(blocks),), 0
+            for block_indices, offset in blocks:
+                syskwargs = {"grid_entry": (count,) ,"grid_shape": out_grid_shape}
+                all_result_oids[block_indices] = self.cm.diag(X.blocks[block_indices].oid, offset, syskwargs=syskwargs)
+                count += 1
+
+            
+            grid_i, grid_j = 0, 0
+            for block_no in diagonal:
+                syskwargs = {"grid_entry": (grid_i,) ,"grid_shape": out_grid_shape}
+                blocks = diagonal[block_no]
+                if len(blocks) == 1:
+                    rarr.blocks[(grid_i, )].oid = all_result_oids[list(blocks.keys())[0]]
+                else:
+                    all_oids = [all_result_oids[key] for key in blocks]
+                    src_params, dst_params,total = array_utils.find_slices(blocks)
```

Comment:
As mentioned in response to your description of your code, I think we can remove `find_slices` and the use of `create_block` in general and just use `concatenate` instead. Do you see any issue with this? This would improve the readability and maintainability of the code.

Reply:
Done. Actually I tried doing using `concetenate` but at that time I didn't do correctly and moved to this solution. But using `concetenate` is definitely better and I have implemented that.